src/libopensc/card-cac.c

```diff
@@ -1008,6 +1094,129 @@ static int cac_decipher(sc_card_t *card,
 	SC_FUNC_RETURN(card->ctx, SC_LOG_DEBUG_VERBOSE, cac_rsa_op(card, data, datalen, out, outlen));
 }
 
+static int parse_properties_object(sc_card_t *card, u8 type,
+    u8 *data, size_t data_len, cac_properties_t *prop, size_t *i)
+{
+	size_t len;
+	u8 *val, *val_end, tag;
+	int parsed = 0;
+
+	if (data_len < 11)
+		return -1;
+
+	/* Initilize: non-PKI applet */
+	prop->privatekey[*i] = 0;
+
+	val = data;
+	val_end = data + data_len;
+	for (; val < val_end; val += len) {
+		/* get the tag and the length */
+		if (sc_simpletlv_read_tag(&val, val_end - val, &tag, &len) != SC_SUCCESS)
+			break;
+
+		switch (tag) {
+		case CAC_TAG_OBJECT_ID:
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			    "TAG: Object ID = 0x%02x 0x%02x", val[0], val[1]);
+			memcpy(&prop->objects[*i], val, 2);
+			parsed++;
+			break;
+
+		case CAC_TAG_BUFFER_PROPERTIES:
+			/* First byte is "Type of Tag Supported" */
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			    "TAG: Buffer Properties: Type of Tag Supported = 0x%02x",
+			    val[0]);
+			prop->simpletlv[*i] = val[0];
+			parsed++;
+			break;
+
+		case CAC_TAG_PKI_PROPERTIES:
+			/* 4th byte is "Private Key Initialized" */
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			    "TAG: PKI Properties: Private Key Initialized = 0x%02x",
+			    val[2]);
+			prop->privatekey[*i] = val[2];
+			parsed++;
+			break;
+
+		default:
+			/* ignore tags we don't understand */
+			sc_debug(card->ctx, SC_LOG_DEBUG_VERBOSE,
+			    "TAG: Unknown (0x%02x)",tag );
+			break;
+		}
+	}
+	if (parsed < 2)
+		return SC_ERROR_INVALID_DATA;
+
+	(*i)++;
```

Comment:
The intention here was to avoid too deep nesting (two for cycles, two switches) and I was not yet sure if I will call the function from one place or from two (since it works with two tags that are handled by this function).

I can refactor the structure so I can pass the whole object that I am filling and increment & check only in one place, while checking for correct bounds.